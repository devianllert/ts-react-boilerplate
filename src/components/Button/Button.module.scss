@import '../../design/design-system';

.button {
  position: relative;

  display: inline-flex;
  align-items: center;
  justify-content: center;

  height: $button-height;
  min-width: $button-min-width;

  margin: $button-margin;
  padding: $button-padding;

  text-align: center;
  vertical-align: middle;

  font-family: $button-font-family;
  font-size: $button-medium-font-size;
  font-weight: $button-font-weight;

  letter-spacing: $button-letter-spacing;

  border: 0;
  border-radius: $button-border-radius;

  color: var(--button-color);
  background-color: var(--brand-primary);

  text-decoration: $button-text-decoration;
  text-transform: $button-text-transform;

  -webkit-tap-highlight-color: transparent;

  outline: 0;

  appearance: none;

  touch-action: manipulation;
  user-select: none;
  cursor: pointer;

  transition: background-color $button-transition;

  &.small {
    font-size: 12px;
    
    height: $button-small-height;
    min-width: $button-small-min-width;
  }

  &.medium {
    font-size: $button-medium-font-size;
    
    height: $button-medium-height;
    min-width: $button-medium-min-width;
  }

  &.large {
    font-size: $button-medium-font-size;
    
    height: $button-large-height;
    min-width: $button-large-min-width;
  }

  &:hover:before {
    opacity: 0.08;
  }

  &:focus:before {
    opacity: 0.24;
  }

  &:active:before {
    opacity: 0.32;
  }

  &:disabled,
  &.disabled {
    background-color: #CDCDCD;
  }
}

@each $variation, $values in $button-color-variations {
  .#{$variation} {
    color: map-get($values, 'color');

    background-color: map-get($values, 'bg');

    &:before {
      content: "";
      position: absolute;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;

      border-radius: inherit;
      
      opacity: 0;

      background-color: #ffffff;

      transition: opacity $button-transition;
    }
  }
}

@each $variation, $values in $button-color-variations {
  .#{$variation}.outlined {
    color: map-get($values, 'bg');

    background-color: transparent;

    border: map-get($values, 'border');

    &:before {
      background-color: currentColor;
    }

    &:disabled,
    &.disabled {
      border-color: #CDCDCD;
    }
  }
}

@each $variation, $values in $button-color-variations {
  .#{$variation}.flat {
    color: map-get($values, 'bg');

    background-color: transparent;

    &:before {
      background-color: currentColor;
    }
  }
}

/* stylelint-disable-next-line no-duplicate-selectors */
.button:disabled,
.button.disabled {
  cursor: not-allowed;

  color: #8E8E8E;

  &:before {
    opacity: 0;
  }
}