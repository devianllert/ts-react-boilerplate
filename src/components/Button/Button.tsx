import React, { ReactChild, ReactElement } from 'react';
import classnames from 'classnames';

import styles from './Button.module.scss';

interface Props {
  disabled?: boolean;
  children: ReactChild | ReactChild[];
}

const Button = (props: Props): ReactElement => {
  const { disabled, children } = props;

  const classes = classnames(
    styles.button,
    disabled && styles.disabled,
  );

  return (
    <button
      className={classes}
      type="button"
      disabled={disabled}
    >
      {children}
    </button>
  );
};

export default Button;
