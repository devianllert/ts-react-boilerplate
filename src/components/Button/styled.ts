import styled from 'styled-components';
import theme, { ThemeSet } from 'styled-theming';

import {
  PRIMARY,
  SUCCESS,
  WARNING,
  DANGER,
} from '../../design/colors';
import { createTransition, duration } from '../../design/transitions';
import { shape } from '../../design/shape';
import { shadows } from '../../design/shadows';


interface BaseButtonProps {
  disabled?: boolean;
  loading?: boolean;
  outlined?: boolean;
  fullWidth?: boolean;
  flat?: boolean;
  appearence?: 'primary' | 'secondary' | 'success' | 'danger' | 'warning';
  size?: 'small' | 'medium' | 'large';
}

const BUTTON_BACKGROUNDS: Record<string, ThemeSet> = {
  primary: PRIMARY,
  success: SUCCESS,
  warning: WARNING,
  danger: DANGER,
};

const DISABLED_COLOR = theme('mode', {
  light: 'rgba(0, 0, 0, 0.3)',
  dark: 'rgba(255, 255, 255, 0.3)',
});

const DISABLED_BACKGROUND_COLOR = theme('mode', {
  light: 'rgba(0, 0, 0, 0.12)',
  dark: 'rgba(255, 255, 255, 0.12)',
});

export const BaseButton = styled.button<BaseButtonProps>`
  position: relative;

  display: inline-flex;
  align-items: center;
  justify-content: center;

  width: ${({ fullWidth }): string => (fullWidth ? '100%' : 'auto')};
  min-width: 88px;
  height: 36px;

  padding: 0 16px;

  text-align: center;
  vertical-align: middle;

  font-family: "Roboto", sans-serif;
  font-size: 14px;
  font-weight: 500;

  letter-spacing: 1px;

  border: 0;
  border-radius: ${shape.borderRadius}px;

  color: #fff;
  background-color: ${({ appearence = 'primary' }): ThemeSet => BUTTON_BACKGROUNDS[appearence]};

  box-shadow: ${shadows[1]};

  text-decoration: none;
  text-transform: uppercase;

  -webkit-tap-highlight-color: transparent;

  outline: 0;

  appearance: none;

  touch-action: manipulation;
  user-select: none;
  cursor: pointer;

  transition: ${createTransition(['background-color', 'box-shadow', 'border'], {
    duration: duration.short,
  })};

  &:before {
    content: '';

    position: absolute;

    top:0;
    left: 0;

    width: 100%;
    height: 100%;

    background-color: transparent;

    transition: ${createTransition('background-color', { duration: duration.short })};

    user-select: none;
    pointer-events: none;
  }

  &:hover:before {
    background-color: rgba(255, 255, 255, 0.12);
  }

  &:focus:before {
    background-color: rgba(255, 255, 255, 0.18);
  }

  &:active:before {
    background-color: rgba(255, 255, 255, 0.24);
  }

  &:disabled {
    color: ${DISABLED_COLOR};
    background-color: ${DISABLED_BACKGROUND_COLOR};

    box-shadow: none;
  }
`;

export const Label = styled.span`
  width: 100%;
  display: inherit;
  align-items: inherit;
  justify-content: inherit;
`;

export const EndIcon = styled.span`
  display: inherit;
  margin-right: -4px;
  margin-left: 8px;
`;

export const StartIcon = styled.span`
  display: inherit;
  margin-right: 8px;
  margin-left: -4px;
`;
