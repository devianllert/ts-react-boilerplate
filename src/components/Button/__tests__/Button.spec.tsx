import React, { ReactChild } from 'react';
import { cleanup, render } from '@testing-library/react';
import { create } from 'react-test-renderer';

import '@testing-library/jest-dom/extend-expect';

import Button from '../Button';

describe('<Button />', (): void => {
  afterEach(cleanup);

  const children: ReactChild = 'Button';

  it('should render a children', (): void => {
    const { getByRole } = render(<Button>{children}</Button>);

    expect(getByRole('button')).toHaveTextContent(children);
  });

  it('should render a disabled button if disabled prop is true', (): void => {
    const { getByRole } = render(<Button disabled>{children}</Button>);

    expect(getByRole('button')).toHaveAttribute('disabled');
  });

  it('should rerender correctly', (): void => {
    const loading = 'Loading...';
    const { rerender, getByRole } = render(<Button>{children}</Button>);

    expect(getByRole('button')).toHaveTextContent(children);

    rerender(<Button disabled>{loading}</Button>);

    expect(getByRole('button')).toHaveTextContent(loading);
    expect(getByRole('button')).toHaveAttribute('disabled');
  });

  it('should render and match the snapshot', (): void => {
    const defaultButton = create(<Button>{children}</Button>);
    const disabledButton = create(<Button disabled>{children}</Button>);

    expect(defaultButton.toJSON()).toMatchSnapshot();
    expect(disabledButton.toJSON()).toMatchSnapshot();
  });
});
