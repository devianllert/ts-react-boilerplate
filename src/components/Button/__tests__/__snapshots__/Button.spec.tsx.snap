// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render and match the snapshot 1`] = `
<button
  className="button"
  type="button"
>
  Button
</button>
`;

exports[`<Button /> should render and match the snapshot 2`] = `
<button
  className="button disabled"
  disabled={true}
  type="button"
>
  Button
</button>
`;
